### ISSUE TITLE
Actual nullability state of closure variables does not affect inference of type arguments

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#33210]
**Version Used**: 

Microsoft Visual Studio Enterprise 2019 Preview
Version 16.0.0 Preview 2.1
VisualStudio.16.Preview/16.0.0-pre.2.1+28529.54
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class C
{
  public static void M1(string s)
  {
    s = "";
    var v = Generic(() => s);
    v.ToString(); // no warnings
  }

  public static void M2(string? s)
  {
    s = "";
    var v = Generic(() => s);
    v.ToString(); // CS8602
  }

  public static extern T Generic<T>(System.Func<T> f);
}
```

**Expected Behavior**:
No warnings `v.ToString()` in `M2`

**Actual Behavior**:
Warning `CS8602 Possible dereference of a null reference.` is reported for `v.ToString()` in `M2`.

**Notes**
Roslyn infers nullability of type arguments based on actual nullability state when there are no closures but fails to do this in the examples above. However when there are closures involved Roslyn seems to fall back to declared nullability of `s` instead of its actual state. E.g. in the following code snippet Roslyn correctly takes the actual nullability state of `s` into account:

```cs
class C
{
  public static void M3(string? s)
  {
    s = "";
    var v = GetT(s);
    v.ToString();
  }

  public static extern T GetT<T>(T t);
}
```

**Comment by gafter** [comment]
This is no longer reproducible.
